src/qibo/backends/clifford.py

```diff
@@ -0,0 +1,679 @@
+from functools import cache
+
+from qibo import gates
+from qibo.backends.numpy import NumpyBackend
+from qibo.config import raise_error
+from qibo.result import CircuitResult
+
+
+class CliffordOperations:
+    """Operations performed by clifford gates on the stabilizers state tableau representation discussed in https://arxiv.org/abs/quant-ph/0406196."""
+
+    def __init__(self):
+        import numpy as np
+
+        self.np = np
+
+    def I(self, tableau, q, nqubits):
+        return tableau
+
+    def H(self, tableau, q, nqubits):
+        new_tab = tableau.copy()
+        self.set_r(
+            new_tab,
+            self.get_r(new_tab, nqubits)
+            ^ (
+                self.get_x(new_tab, nqubits)[:, q] * self.get_z(new_tab, nqubits)[:, q]
+            ).flatten(),
+        )
+        new_tab[:, [q, nqubits + q]] = new_tab[:, [nqubits + q, q]]
+        return new_tab
+
+    def CNOT(self, tableau, control_q, target_q, nqubits):
+        new_tab = tableau.copy()
+        x, z = self.get_x(new_tab, nqubits), self.get_z(new_tab, nqubits)
+        self.set_r(
+            new_tab,
+            self.get_r(new_tab, nqubits)
+            ^ (x[:, control_q] * z[:, target_q]).flatten()
+            * (x[:, target_q] ^ z[:, control_q] ^ 1).flatten(),
+        )
+
+        new_tab[:-1, target_q] = x[:, target_q] ^ x[:, control_q]
+        new_tab[:-1, nqubits + control_q] = z[:, control_q] ^ z[:, target_q]
+        return new_tab
+
+    def CZ(self, tableau, control_q, target_q, nqubits):
+        """Decomposition --> H-CNOT-H"""
+        new_tab = tableau.copy()
+        r, x, z = (
+            self.get_r(new_tab, nqubits),
+            self.get_x(new_tab, nqubits),
+            self.get_z(new_tab, nqubits),
+        )
+        self.set_r(
+            new_tab,
+            r
+            ^ (x[:, target_q] * z[:, target_q]).flatten()
+            ^ (
+                x[:, control_q]
+                * x[:, target_q]
+                * (z[:, target_q] ^ z[:, control_q] ^ 1)
+            ).flatten()
+            ^ (x[:, target_q] * (z[:, target_q] ^ x[:, control_q])).flatten(),
+        )
+        new_tab[:-1, [nqubits + control_q, nqubits + target_q]] = self.np.vstack(
+            (x[:, target_q] ^ z[:, control_q], z[:, target_q] ^ x[:, control_q])
+        ).T
+        return new_tab
+
+    def S(self, tableau, q, nqubits):
+        new_tab = tableau.copy()
+        x, z = self.get_x(new_tab, nqubits), self.get_z(new_tab, nqubits)
+        self.set_r(
+            new_tab,
+            self.get_r(new_tab, nqubits) ^ (x[:, q] * z[:, q]).flatten(),
+        )
+        new_tab[:-1, nqubits + q] = z[:, q] ^ x[:, q]
+        return new_tab
+
+    def Z(self, tableau, q, nqubits):
+        """Decomposition --> S-S"""
+        new_tab = tableau.copy()
+        r, x, z = (
+            self.get_r(new_tab, nqubits),
+            self.get_x(new_tab, nqubits),
+            self.get_z(new_tab, nqubits),
+        )
```

Comment:
This pattern is repeated over and over. Could you factor out as a separate function?